docs/a2a-part-of-indy-sdk-2.0/README.md

```diff
@@ -0,0 +1,656 @@
+
+##### CredentialOfferCreated
+Issuer specific state that means that Credential Offer message has been created and sent to Prover.
+Transitions:
+* CredentialRequestReceived - prover replied with Credential Request message.
+* CredentialRejectReceived - prover replied woth Credential Reject message.
+    
+##### CredentialRequestReceived
+Issuer specific state that means that Prover replied with Credential Request message on offer.
+Transitions:
+* CredentialCreated - credential created and sent to prover
+* CredentialRejectCreated - issuer interrupted credential issuance process for some reason.
+
+##### CredentialCreated
+Issuer specific state that means that Credential message has been created and sent to Prover.
+Transitions:
+* CredentialAckReceived - prover accepted credential and replied with Credential Ack message.
+* CredentialRejectReceived - prover rejected credential for some reason.
+    
+##### CredentialAckReceived
+Issuer specific state that means that Prover replied with Credential Ack message on credential.
+Transitions:
+* Finished - credential issuance process is completed.
+
+##### CredentialOfferReceived
+Prover specific state that means that Credential Offer message has been received from Issuer.
+Transitions:
+* CredentialRequestCreated - prover accepted offer and replied with Credential Request message.
+* CredentialRejectCreated - prover rejected credential offer for some reason.
+
+##### CredentialRequestCreated
+Prover specific state that means that Credential Offer message has been accepted and 
+Credential Request message created and sent to Issuer.
+Transitions:
+* CredentialReceived - credential message received from Issuer.
+* CredentialRejectReceived - issuer interrupted issuance process for some reason.
+
+##### CredentialReceived
+Prover specific state that means that Credential message has been received from Issuer.
+Transitions:
+* CredentialAckCreated - prover accepted credential and replied with Credential Ack message.
+* CredentialRejectCreated - prover rejected credential and interrupted issuance process for some reason.
+
+##### CredentialAckCreated
+Prover specific state that means that Credential Ack message created and sent to Issuer.
+Transitions:
+* Finished - credential issuance process is completed.
+
+##### CredentialRejectCreated
+Common state for both parties. Indicates interruption of issuance process.
+Transitions:
+* Finished - credential issuance process is interrupted.
+
+##### CredentialRejectReceived
+Common state for both parties. Indicates interruption of issuance process by opposite side.
+Transitions:
+* Finished - credential issuance process is interrupted.
+
+#### Choreography Diagram
+
+![Choreography Diagram](./credential-issuance.png)
+
+#### Workflow
+
+The Credential Issuance positive workflow is demonstrated on the following diagram:
+
+![Workflow](credential-issuance-positive-flow.svg)
```

Comment:
Unless I am missing something, the way IndySDK is currently designed requires prover to receive credential offer, before credential request can be submitted by prover. It seems from the diagram like this restriction will stay. Why?
If I have connections with an institution, it seems natural that I should be able to request something from them. 

In my organization we've implemented credential exchange prototype between mobile app/backend on top of IndySDK without VCX. We basically ended up with equivalent of something like "request credential offer" message type. Did we miss something? 
Aditionally we've also implemented some API on issuer side which enabled the receiving side to ask issuer what kind of credentials can he issue. So that was typically the API the credential requester called first. Once he knew what's there to be offered, he would send "request credential offer" message asking for a kind of credential. The issuer decided whether the requester is "eligible" to get such credential based some non-indy application logic/data.
Did we not "get" something about IndySDK design? Is/will be flow of an agent directly requesting other agent for specific type of credential supported?

Reply:
@Patrik-Stas There is an attack that can be performed by a credential issuer if they do not present an offer. It allows them to embed data into a credential that the requester does not know about.

Reply:
@brentzundel or @mikelodder7 or @lovesh may be able to clarify.

Reply:
@dhh1128 I see, fair enough if this kind of 3 step credential exchange protocol is security requirement. However, what about the functional requirement? Having the ability to ask someone to give you a credential of certain type?
Wouldn't it make sense to have something like "credential offer request" as a standard message type? (sounds awkward, a better name could be given).

Reply:
@Patrik-Stas The need for sending credential offer is for the issuer to send a nonce to ensure that the holder does know the value of the committed attributes. One such attribute is link secret. The holder never reveals the link secret, he only sends a commitment to the issuer. The holder uses the nonce sent as part of credential offer to send a proof (prior to issuance) to the issuer that he does know the link secret he is committing to. The nonce does not have to sent by the issuer always, the prover can select a nonce from another source like the current merkle root of the ledger or a randomness beacon.

> what about the functional requirement? Having the ability to ask someone to give you a credential of certain type?

Yes, we can introduce a message by holder indicating the "type" of credential it needs.

Reply:
The important part is the holder shouldn't be allowed to generate the nonce otherwise cheating is possible. Selecting a nonce like the merkle root that the issuer can verify is acceptable. Otherwise, like @lovesh says you need to have the nonce come from the issuer.